### PR TITLE
pageserver: do fewer heatmap uploads for tiny tenants

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Currently we do a large number of heatmap uploads for tiny tenants.  "tiny" in this context is defined as being less than a single layer in size.  These uploads are triggered by atime changes rather than changes in the set of layers.

Uploading heatmaps for atime changes on small tenants isn't useful, because even without bumping these atimes, disk usage eviction still avoids evicting the largest resident layer of a tenant, which in practice keeps tiny/empty tenants mostly resident irrespective of atimes.

## Summary of changes

- For tenants smaller than one checkpoint interval, only upload heatmap if the set of layers has changed, not if only the atimes have changed.
- Include the heatmap period in the uploaded heatmap, as a precursor to implementing https://github.com/neondatabase/neon/issues/6200 (auto-adjusting download intervals to match upload intervals)

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/bin/pageserver.rs [hunk 1]
```diff
@@ -383,7 +383,7 @@ fn start_pageserver(
     let shutdown_pageserver = tokio_util::sync::CancellationToken::new();
 
     // Set up remote storage client
-    let remote_storage = create_remote_storage_client(conf)?;
+    let remote_storage = Some(create_remote_storage_client(conf)?);
```
Comment (petuhovskiy) [inline pageserver/src/bin/pageserver.rs:386]: This is probably unrelated change from another PR: https://github.com/neondatabase/neon/pull/7722/files

File: pageserver/src/tenant/secondary/heatmap_uploader.rs [hunk 2]
```diff
@@ -396,15 +407,31 @@ async fn upload_tenant_heatmap(
...
+    // layers from changes only in atimes.
+    let heatmap_size_bytes = heatmap.get_stats().bytes;
+    let layers_only_bytes =
+        serde_json::to_vec(&heatmap.strip_atimes()).map_err(|e| anyhow::anyhow!(e))?;
+    let layers_only_digest = md5::compute(&layers_only_bytes);
+    if heatmap_size_bytes < tenant.get_checkpoint_distance() {
+        // For small tenants, skip upload if only atimes changed. This avoids doing frequent
+        // uploads from long-idle tenants whose atimes are just incremented by periodic
+        // size calculations.
+        if Some(&layers_only_digest) == last_upload.as_ref().map(|d| &d.layers_only_digest) {
```
Comment (petuhovskiy) [inline pageserver/src/tenant/secondary/heatmap_uploader.rs:427]: Do we have to upload heatmap when layers change? What bad things can happen if we stop uploading heatmaps?

File: pageserver/src/tenant/secondary/heatmap_uploader.rs [hunk 3]
```diff
@@ -347,12 +344,25 @@ enum UploadHeatmapError {
...
+    // Digest of json-encoded HeatMapTenant
+    uploaded_digest: md5::Digest,
+
+    // Digest without atimes set.
+    layers_only_digest: md5::Digest,
+}
+
 /// The inner upload operation.  This will skip if `last_digest` is Some and matches the digest
 /// of the object we would have uploaded.
 async fn upload_tenant_heatmap(
```
Comment (petuhovskiy) [inline pageserver/src/tenant/secondary/heatmap_uploader.rs:362]: It is true that each tenant uploads new heatmap every 60 seconds by default? Isn't it causing too much load?

File: pageserver/src/tenant/secondary/heatmap.rs [hunk 4]
```diff
@@ -15,6 +15,14 @@ pub(super) struct HeatMapTenant {
...
 
     pub(super) timelines: Vec<HeatMapTimeline>,
+
+    /// Uploaders provide their own upload period in the heatmap, as a hint to downloaders
+    /// of how frequently it is worthwhile to check for updates.
+    ///
+    /// This is optional for backward compat, and because we sometimes might upload
+    /// a heatmap explicitly via API for a tenant that has no periodic upload configured.
+    #[serde(default)]
+    pub(super) upload_period_ms: Option<u128>,
```
Comment (petuhovskiy) [inline pageserver/src/tenant/secondary/heatmap.rs:25]: Change to heatmap structure will change the content bytes and every tenant will have to reupload the new version. Will it cause too much load?

**Review by petuhovskiy — APPROVED** [review]

File: pageserver/src/tenant/secondary/heatmap.rs (hunk 4 above)
Comment (jcsp) [reply pageserver/src/tenant/secondary/heatmap.rs:25]: We already do uploads of all heatmaps after a restart (jittered across the heatmap period). 
 `heatmap_upload_concurrency` (default 8) limits this to prevent overwhelming anything.

File: pageserver/src/tenant/secondary/heatmap_uploader.rs (hunk 3 above)
Comment (jcsp) [reply pageserver/src/tenant/secondary/heatmap_uploader.rs:362]: The default is 60s in the storage controller, but currently that only manages a smaller number of large tenants.  For the general population of tiny tenants, a 5 minute period is used (set in control plane).  `heatmap_upload_concurrency` also reduces risk of this interfering with any other work.

File: pageserver/src/tenant/secondary/heatmap_uploader.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/secondary/heatmap_uploader.rs:427]: Functionally, nothing breaks -- heatmaps+secondaries are an optimization to make migrations faster.  In practice, an out-of-date heatmap results in a less-warm state on a secondary location, so there is more work to do when we need to migrate a tenant (e.g. after a failure).  Across large numbers of tenants, if each one is missing a few layers then the aggregate amount of work to recover from a pageserver failure can become very large (100s of GB to download)
